drivers: spi: Added MEC172x full duplex qmspi driver version

jvasanth1 opened the pull request:
Add MEC172x qmspi driver version 2 to support full duplex transfers as expected by the Zephyr spi driver model.
On every spi clock we transmit one bit and receive one bit. This driver will work with Zephyr SPI NOR driver.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
drivers: spi: Added MEC172x full duplex qmspi driver

Add MEC172x full duplex qmspi driver version to support full
duplex transfers as expected by the Zephyr spi driver model.
On every spi clock we transmit one bit and receive one bit.
This driver will work with Zephyr SPI NOR driver.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

mbolivar-nordic commented on drivers/spi/Kconfig.xec_qmspi_v2:
```diff
@@ -0,0 +1,12 @@
+
+config SPI_XEC_QMSPI_V2
+	bool "Microchip XEC QMSPI V2 driver"
+	default n
```
Above, please add

```
DT_COMPAT_MICROCHIP_XEC_QMSPI_V2 := microchip,xec-qmspi-v2
```

then, here:

```
	default $(dt_compat_enabled,$(DT_COMPAT_MICROCHPI_XEC_QMSPI_V2))
```

This is generally the right thing to do for kconfig values. Saying `default n` is always redundant and makes things harder to use for your users, because then they have to enable CONFIG_SPI **and** your IP block specific Kconfig option.

jvasanth1 replied:
We have two SPI drivers for MEC172x, the latest version 2 was added to work with the Zephyr NOR driver. 
So the application can enable only one of the two XEC SPI drivers, CONFIG_SPI_XEC_QMSPI_LDMA or CONFIG_SPI_XEC_QMSPI_V2.

mbolivar-nordic commented on drivers/spi/spi_xec_qmspi_v2.c:
```diff
@@ -0,0 +1,746 @@
+	case SPI_LINES_SINGLE:
+		qlines = XEC_QSPI_C_IFC_1X;
+		break;
+#if DT_INST_PROP(0, lines) > 1
+	case SPI_LINES_DUAL:
+		qlines = XEC_QSPI_C_IFC_2X;
+		break;
+#endif
+#if DT_INST_PROP(0, lines) > 2
+	case SPI_LINES_QUAD:
+		qlines = XEC_QSPI_C_IFC_4X;
+		break;
+#endif
```
What's going on here? It looks like the `lines` value should be saved in per-instance configuration structure members instead. Hard-coding an instance number like this in a multi-instance driver is almost guaranteed to be incorrect.

mbolivar-nordic commented on dts/bindings/spi/microchip,xec-qmspi-v2.yaml:
```diff
@@ -0,0 +1,112 @@
+
+    lines:
+      type: int
+      required: false
+      description: |
+        QMSPI data lines 1, 2, or 4. 1 data line is full-duplex
+        MOSI and MISO or half-duplex on MOSI only. Lines set to 2
+        or 4 indicate dual or quad I/O modes. Defaults to 1.
```
Please use an `enum` and document the reason for this default.

jvasanth1 replied:
using `enum`

mbolivar-nordic commented on soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz:
```diff
@@ -47,4 +47,8 @@ config PECI_XEC
 	default y
 	depends on PECI
 
+config SPI_XEC_QMSPI_V2
+	default n
+	depends on SPI
+
```
What's going on here?

jvasanth1 replied:
We have two SPI drivers for MEC172x, the latest version 2 was added to work with the Zephyr NOR driver.
So the application can enable only one of the two XEC SPI drivers, CONFIG_SPI_XEC_QMSPI_LDMA or CONFIG_SPI_XEC_QMSPI_V2.
We cannot turn on both (qmspi_v2) and qmspi_ldma at the same time.

jackrosenthal replied:
Should this be a `choice` then?

jvasanth1 replied:
yes, will consider `choice`

jvasanth1 replied:
updated with choice

mbolivar-nordic commented on drivers/spi/spi_xec_qmspi_v2.c:
```diff
@@ -0,0 +1,746 @@
+		.irq_pri = DT_INST_IRQ(i, priority),			\
+		.pcr_idx = DT_INST_PROP_BY_IDX(i, pcrs, 0),		\
+		.pcr_pos = DT_INST_PROP_BY_IDX(i, pcrs, 1),		\
+		.port_sel = DT_INST_PROP_OR(i, port-sel, 0),		\
```
This must be unused (it won't compile). What's going on here?

Please double check that all config members are used and remove any unnecessary properties.

jvasanth1 replied:
will remove unused properties

jvasanth1 replied:
removed unused properties

mbolivar-nordic commented on drivers/spi/spi_xec_qmspi_v2.c:
```diff
@@ -0,0 +1,746 @@
+		.pcr_idx = DT_INST_PROP_BY_IDX(i, pcrs, 0),		\
+		.pcr_pos = DT_INST_PROP_BY_IDX(i, pcrs, 1),		\
+		.port_sel = DT_INST_PROP_OR(i, port-sel, 0),		\
+		.chip_sel = DT_INST_PROP_OR(i, chip-select, 0),		\
```
This is not correct either; what's going on?

jvasanth1 replied:
chip_sel is being used in the driver.

jvasanth1 replied:
updated accordingly

mbolivar-nordic commented on drivers/spi/spi_xec_qmspi_v2.c:
```diff
@@ -0,0 +1,746 @@
+	};								\
+	static const struct spi_xec_qspi_config xec_qspi_config_##i = {	\
+		.regs = (struct qspi_regs *) DT_INST_REG_ADDR(i),	\
+		.cs1_freq = DT_INST_PROP_OR(i, cs1-freq, 0),		\
```
Same here.

jvasanth1 replied:
cs1_freq is being used in the driver.

jvasanth1 replied:
updated to `cs1_freq`

mbolivar-nordic reviewed (dismissed):
There are several issues with the binding and the driver's usage of it. How was this tested?

jvasanth1 commented:
Hi @mbolivar-nordic could you help to review my responses.

jackrosenthal commented on soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz:
```diff
@@ -47,4 +47,8 @@ config PECI_XEC
 	depends on PECI
 
+config SPI_XEC_QMSPI_V2
+	default n
```
`default n` is the default value for booleans. No need to write it out.

jvasanth1 replied:
using choice now

jackrosenthal commented on drivers/spi/Kconfig.xec_qmspi_v2:
```diff
@@ -0,0 +1,12 @@
+
+config SPI_XEC_QMSPI_V2
+	bool "Microchip XEC QMSPI V2 driver"
+	default n
```
`default n` is the default for booleans. No need to write it out.

jvasanth1 replied:
using choice now

jackrosenthal commented on drivers/spi/Kconfig.xec_qmspi_v2:
```diff
@@ -0,0 +1,12 @@
+	depends on SOC_SERIES_MEC172X
+	help
+	  Enable support for the Microchip XEC QMSPI driver
+	  for MEC172x.
```
This could use an explanation as to what "V2" means? Why would someone want the non-V2 driver?

jvasanth1 replied:
updated in Kconfig.xec_qmspi_v2_ldma

fabiobaltieri commented:
@jvasanth1 current version of this PR ([6728f12](https://github.com/zephyrproject-rtos/zephyr/pull/46273/commits/6728f12c90271ce6ee839a58844d0ada44a7f272)) does not build, maybe some changes have been left out?

jvasanth1 commented:
> @jvasanth1 current version of this PR ([6728f12](https://github.com/zephyrproject-rtos/zephyr/pull/46273/commits/6728f12c90271ce6ee839a58844d0ada44a7f272)) does not build, maybe some changes have been left out?

can you check now?

jackrosenthal commented on soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz:
```diff
@@ -39,12 +39,22 @@ config ESPI_XEC_V2
+	config SPI_XEC_QMSPI_LDMA
+		bool "Use QMSPI driver with LDMA (local dma)"
+
+	config SPI_XEC_QMSPI_V2
```
Here too

jvasanth1 replied:
updated in Kconfig.xec_qmspi_v2_ldma

jvasanth1 commented:
can anybody suggest why few checks are failing?

fabiobaltieri commented:
> can anybody suggest why few checks are failing?

The error from https://github.com/zephyrproject-rtos/zephyr/runs/7415641683?check_suite_focus=true is:

```
FAILED: zephyr/drivers/spi/CMakeFiles/drivers__spi.dir/spi_xec_qmspi_ldma.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DCORE_CM4 -DHSE_VALUE=16000000 -DKERNEL -DSTM32F412Cx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../../../../../../../include/zephyr -I../../../../../../../include -Izephyr/include/generated -I../../../../../../../soc/arm/st_stm32/stm32f4 -I../../../../../../../drivers -I../../../../../../../soc/arm/st_stm32/common -I/__w/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/__w/zephyr/modules/hal/st/sensor/vl53l0x/api/core/inc -I../../../../../../../drivers/sensor/vl53l0x -I/__w/zephyr/modules/hal/st/sensor/stmemsc/hts221_STdC/driver -I/__w/zephyr/modules/hal/stm32/stm32cube/stm32f4xx/soc -I/__w/zephyr/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include -I/__w/zephyr/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include/Legacy -I/__w/zephyr/modules/hal/stm32/stm32cube/common_ll/include -isystem ../../../../../../../lib/libc/minimal/include -isystem /opt/toolchains/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include -isystem /opt/toolchains/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include-fixed -Os -imacros /__w/zephyr/zephyr/twister-out/96b_argonkey/samples/boards/96b_argonkey/sensors/sample.board.96b_argonkey.sensors/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/opt/toolchains/zephyr-sdk-0.14.2/arm-zephyr-eabi/arm-zephyr-eabi -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/samples/boards/96b_argonkey/sensors=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/spi/CMakeFiles/drivers__spi.dir/spi_xec_qmspi_ldma.c.obj -MF zephyr/drivers/spi/CMakeFiles/drivers__spi.dir/spi_xec_qmspi_ldma.c.obj.d -o zephyr/drivers/spi/CMakeFiles/drivers__spi.dir/spi_xec_qmspi_ldma.c.obj -c /__w/zephyr/zephyr/drivers/spi/spi_xec_qmspi_ldma.c
/__w/zephyr/zephyr/drivers/spi/spi_xec_qmspi_ldma.c: In function 'qmspi_reset':
/__w/zephyr/zephyr/drivers/spi/spi_xec_qmspi_ldma.c:140:16: error: invalid use of undefined type 'struct qmspi_regs'
  140 |  taps[0] = regs->TM_TAPS;
      |                ^~
```

which means that the compiler has no clue what `struct qmspi_regs`. That's define in the soc headers, but if you notice the error is from this test:

```
INFO    - 604/608 96b_argonkey              samples/boards/96b_argonkey/sensors/sample.board.96b_argonkey.sensors FAILED Build failure (build)
```

which has nothing to do with the MEC172x, so it's trying to compile a driver for a platform that has nothing to do with it. This is because when moving from `Kconfig.xec_qmspi_ldma` to `Kconfig.xec_qmspi_v2_ldma` you dropped the `depends on SOC_SERIES_MEC172X` bit that prevents the driver from being built on other platforms.

jvasanth1 commented:
thanks @fabiobaltieri 
I fixed the errors by adding the SOC_SERIES_MEC172X depends on

jvasanth1 commented on soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz:
```diff
@@ -39,12 +39,22 @@ config ESPI_XEC_V2
+	help
+	 Select Microchip XEC QMSPI driver implementation.
+
+	config SPI_XEC_QMSPI_LDMA
```
updated in Kconfig.xec_qmspi_v2_ldma

jackrosenthal commented on drivers/spi/Kconfig.xec_qmspi_v2_ldma:
```diff
@@ -0,0 +1,18 @@
+config SPI_XEC_QMSPI_LDMA
+	bool "Use QMSPI driver with LDMA (local dma)"
+
+config SPI_XEC_QMSPI_V2
```
where is v1? why would someone want to use v1 instead of v2?

should we be deleting v1?

jvasanth1 replied:
v1 is the ldma (local dma based) qmspi driver, we are not deleting it. 
The v2 driver supports only full duplex to work with the Zephyr NOR driver.

jackrosenthal replied:
should it not be called v2 then? based on the name, it sounds like v2 replaces v1, but you say it does not?

jvasanth1 replied:
it replaces it only for a particular use case - to work with Zephyr NOR driver, because the ldma qmspi driver is incompatible with Zephyr NOR.
do you think we should call it differently?

jvasanth1 replied:
we are thinking to change to to `SPI_XEC_QMSPI_SIMPLE` (or `SPI_XEC_QMSPI_FD`) as it does only full duplex. Can you help to confirm?

jackrosenthal replied:
`SPI_XEC_QMSPI_FULL_DUPLEX` would avoid uncommon abbreviations.

jvasanth1 replied:
Thanks!

jvasanth1 replied:
updated to use SPI_XEC_QMSPI_FULL_DUPLEX

jvasanth1 commented:
@mbolivar-nordic can you help to review the responses to your review comments.

jvasanth1 commented on drivers/spi/Kconfig:
```diff
@@ -97,10 +97,10 @@ source "drivers/spi/Kconfig.npcx_fiu"
 
 source "drivers/spi/Kconfig.mchp_mss_qspi"
 
+source "drivers/spi/Kconfig.xec_qmspi_v2_ldma"
```
`kconfig,xec_qmspi` is for MEC15xx version of qmspi driver `spi_xec_qmspi.c`

galak commented on drivers/spi/Kconfig.xec_qmspi_v2_ldma:
```diff
@@ -0,0 +1,18 @@
+	 Select Microchip XEC QMSPI driver implementation.
+
+config SPI_XEC_QMSPI_LDMA
+	bool "Use QMSPI driver with LDMA (local dma)"
```
add:
```
default y
depends on DT_HAS_MICROCHIP_XEC_QMSPI_LDMA_ENABLED
```

jvasanth1 replied:
replaced with choice in spi_xec_qmspi_full_duplex.h
could you check that?

galak commented on drivers/spi/Kconfig.xec_qmspi_v2_ldma:
```diff
@@ -0,0 +1,18 @@
+	bool "Use QMSPI driver with LDMA (local dma)"
+
+config SPI_XEC_QMSPI_V2
+	bool "Use QMSPI V2 driver to work with Zephyr NOR driver"
```
add
```
default y
depends on DT_HAS_MICROCHIP_XEC_QMSPI_V2
```

jvasanth1 replied:
replaced with choice in spi_xec_qmspi_full_duplex.h
could you check that?

galak commented on drivers/spi/Kconfig.xec_qmspi_v2_ldma:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2022 Microchip Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+choice
```
Why the choice?  Isn't this determined by what nodes are in devicetree with which compatibles?

jvasanth1 replied:
the choice was added based on review comments in this PR.
we have two flavor of driver for MEC172x qmspi and depending on the application use case we want to have a choice to pick the right version.

galak reviewed (changes requested):
Some Kconfig updates.

jvasanth1 commented:
@albertofloyd could you help to review and approve?

albertofloyd commented on drivers/spi/spi_xec_qmspi_full_duplex.c:
```diff
@@ -0,0 +1,710 @@
+ */
+static void xec_qspi_reset(struct qspi_regs *regs)
+{
+	uint32_t taps[3];
```
magic number

jvasanth1 replied:
added macro

albertofloyd commented on drivers/spi/spi_xec_qmspi_full_duplex.c:
```diff
@@ -0,0 +1,710 @@
+			>> XEC_QSPI_M_CLK_DIV_POS;
+
+	if (fdiv == 0) {
+		fdiv = 256u;
```
magic number

jvasanth1 replied:
updated with correct macro

albertofloyd commented on drivers/spi/spi_xec_qmspi_full_duplex.c:
```diff
@@ -0,0 +1,710 @@
+	uint32_t m;
+
+	ptbl = smode_tbl;
+	if (qspi_get_frequency(regs) >= 48000000u) {
```
magic number, can this use any clock control api or macros?

albertofloyd replied:
i.e. MEC152X_QSPI_SRC_CLOCK_HZ

jvasanth1 replied:
will change it to MHZ(48)
it has no reference to MEC15xx or MEC17xx specific

albertofloyd commented on drivers/spi/spi_xec_qmspi_full_duplex.c:
```diff
@@ -0,0 +1,710 @@
+ *  Mode 3: CPOL=1 CPHA=1 (CHPA_MISO=0 and CHPA_MOSI=1)
+ */
+
+const uint8_t smode_tbl[4] = {
```
Consider adding few lines as documentation for these values

jvasanth1 replied:
added documentation for the byte values

albertofloyd commented on drivers/spi/spi_xec_qmspi_full_duplex.c:
```diff
@@ -0,0 +1,710 @@
+		ptbl = smode48_tbl;
+	}
+
+	m = (uint32_t)ptbl[smode & 0x03];
```
macro?

jvasanth1 replied:
address mask for lower two bits, do we need macro?
should I change to (BIT(0) | BIT(1)) ?

jvasanth1 replied:
added macro

albertofloyd commented on drivers/spi/spi_xec_qmspi_full_duplex.c:
```diff
@@ -0,0 +1,710 @@
+	qspi_set_frequency(regs, spi_conf->frequency);
+
+	smode = 0;
+	if ((spi_conf->operation & SPI_MODE_CPHA) != 0U) {
```
Macros?

jvasanth1 replied:
added macros BIT(0), BIT(1)

albertofloyd replied:
Actually meant adding something like 
   #define SMODE_CPOL  BIT(1)

jvasanth1 commented:
@albertofloyd could you check the fixes

jvasanth1 commented on dts/bindings/spi/microchip,xec-qmspi-v2.yaml:
```diff
@@ -0,0 +1,117 @@
+# Copyright (c) 2018, Google LLC.
+# Copyright (c) 2021, Microchip Technology Inc.
```
fixed

jvasanth1 commented on drivers/spi/spi_xec_qmspi_v2.c:
```diff
@@ -0,0 +1,710 @@
+/* #define XEC_QMSPI_DEBUG */
+#ifdef XEC_QMSPI_DEBUG
+#include <sys/printk.h>
+#endif
```
removed it

fabiobaltieri commented on drivers/spi/Kconfig.xec_qmspi_v2:
```diff
@@ -0,0 +1,18 @@
+choice
+	prompt "Select Microchip XEC QMSPI driver"
+	default SPI_XEC_QMSPI_LDMA
+	depends on SPI && SOC_SERIES_MEC172X
```
`SPI` here should be redundant since this is already under drivers/spi. `SOC_SERIES_MEC172X` could also probably go and instead I think you could use something like

`depends on DT_HAS_MICROCHIP_XEC_QMSPI_FULL_DUPLEX || DT_HAS_MICROCHIP_XEC_QMSPI_LDMA`

jvasanth1 replied:
updated to
`depends on DT_HAS_MICROCHIP_XEC_QMSPI_FULL_DUPLEX_ENABLED || DT_HAS_MICROCHIP_XEC_QMSPI_LDMA_ENABLED`

MaureenHelm commented on drivers/spi/spi_xec_qmspi_full_duplex.c:
```diff
@@ -0,0 +1,715 @@
+		.pcr_idx = DT_INST_PROP_BY_IDX(i, pcrs, 0),		\
+		.pcr_pos = DT_INST_PROP_BY_IDX(i, pcrs, 1),		\
+		.chip_sel = DT_INST_PROP_OR(i, chip_select, 0),		\
+		.width = DT_INST_PROP_OR(0, lines, 1),			\
```
```suggestion
		.width = DT_INST_PROP_OR(i, lines, 1),			\
```

jvasanth1 replied:
fixed

MaureenHelm commented on dts/bindings/spi/microchip,xec-qmspi-full-duplex.yaml:
```diff
@@ -0,0 +1,117 @@
+
+compatible: "microchip,xec-qmspi-full-duplex"
+
+include: [spi-controller.yaml, pinctrl-device.yaml]
```
Consider extracting common properties from `microchip,xec-qmspi-full-duplex` and `microchip,xec-qmspi-ldma` into a shared include

jvasanth1 replied:
will consider in a separate PR later, will like to get this PR merged soon

MaureenHelm replied:
We shouldn't introduce a new file that's almost identical to one that already exists in the tree:
```diff
$ diff microchip,xec-qmspi-ldma.yaml microchip,xec-qmspi-full-duplex.yaml 
2c2
< # Copyright (c) 2021, Microchip Technology Inc.
---
> # Copyright (c) 2022, Microchip Technology Inc.
7c7
< compatible: "microchip,xec-qmspi-ldma"
---
> compatible: "microchip,xec-qmspi-full-duplex"
43c43,48
<         or 4 indicate dual or quad I/O modes. Defaults to 1.
---
>         or 4 indicate dual or quad I/O modes.
>         Defaults to 1 for current driver's support for full-duplex spi.
>       enum:
>         - 1
>         - 2
>         - 4
45c50
<     port_sel:
---
>     port-sel:
54c59
<     chip_select:
---
>     chip-select:
```

Can you go ahead and fix this while the merge window is closed?

jvasanth1 replied:
yes, working on this

jvasanth1 replied:
moved all common properties to a shared include; added as a separate commit

MaureenHelm replied:
The shared include commit needs to be first in the series.

jvasanth1 replied:
there are only 2 commits in the PR, shall I merge the shared include commit with previous commit?

jvasanth1 replied:
also with only the ldma driver (prior to this PR), there is no other driver to compare for moving common properties to a separate include.

jvasanth1 replied:
moved the shared include commit to be first in the series.

MaureenHelm reviewed (commented):
Does this driver build in CI? (e.g., is this board enabled for `tests/drivers/spi/`?)

@galak @jackrosenthal @mbolivar-nordic please revisit

jvasanth1 commented:
Hi @MaureenHelm can you approve and merge this PR?

jvasanth1 committed:
dts: microchip: spi: move common properties to a separate include

updated dts variables naming (hyphen instead of underscore).
moved all properties in microchip ldma yaml to a separate include file,
these properties will be common with the (to be added)
full duplex spi driver.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 commented:
@albertofloyd can you help to approve

MaureenHelm commented:
@galak @mbolivar-nordic please revisit

galak commented on drivers/spi/Kconfig.xec_qmspi_v2:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2022 Microchip Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+choice
```
This seems to be 2 distrnict drivers so `choice` doesn't make sense.

Should just be something like:

```
config SPI_XEC_QMSPI_LDMA
   prompt "QMSPI driver with LDMA (local dma)"
   default y
   depends on DT_HAS_MICROCHIP_XEC_QMSPI_LDMA_ENABLED
   ... 

config SPI_XEC_QMSPI_FULL_DUPLEX
   prompt "Use QMSPI full duplex driver to work with Zephyr NOR driver"
   default y
   depends on DT_HAS_MICROCHIP_XEC_QMSPI_FULL_DUPLEX_ENABLED
   ...

jvasanth1 replied:
choice is required because only one driver can work at a time, both work with the same hardware

see discussion here https://github.com/zephyrproject-rtos/zephyr/pull/46273#discussion_r896078492

jvasanth1 replied:
this driver was added to support full duplex transfers as expected by the Zephyr spi driver model to work with Zephyr SPI NOR driver. This cannot be added with existing driver as a feature.

galak replied:
> This cannot be added with existing driver as a feature.

Why?

Seems like there is at least common defines, structs, etc that should NOT be duplicated.

jvasanth1 replied:
This full duplex driver is compatible with SPI NOR driver which expects on every spi clock we transmit one bit and receive one bit. Here for every transmit buffer you need a receive buffer.

The ldma driver supports quad and dual mode transfers using local dma and uses a model where it processes all transmit buffers and then process all receive buffers, if any.

Merging is complex.

This full duplex driver is required for our Zephyr chromium integration which is on hold due to this PR since June.

albertofloyd replied:
@galak we also consider the same request that you propose but due the low level operations between the 2 drivers, the result will be a merged one plagued with ifdefs everywhere which would be hard to maintain and debug.

galak replied:
> @galak we also consider the same request that you propose but due the low level operations between the 2 drivers, the result will be a merged one plagued with ifdefs everywhere which would be hard to maintain and debug.

I don't understand this comment.  If its the same hardware, than the structs, defines, etc that are being used would be the same.

jvasanth1 replied:
we had discussion with @MaureenHelm and agreed to keep the 2 different drivers but update the full_duplex driver to avoid the duplicate `qspi_regs` sturcture

jvasanth1 replied:
updated to use the common qmspi_regs structure

jvasanth1 replied:
also cleaned up the Kconfig

galak reviewed (dismissed):
Kconfig needs cleaning up.

jvasanth1 commented:
Hi @MaureenHelm and @galak 
could you help to review and approve this PR?

jvasanth1 commented:
@albertofloyd @fabiobaltieri 
can you help to re-approve?

jvasanth1 commented:
can we merge this PR?
